src/VisualStudio/Core/Def/Implementation/Diagnostics/SolutionCrawlerEventHandlerWithDelay.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
+{
+    internal sealed partial class TaskCenterSolutionAnalysisProgressReporter
+    {
+        /// <summary>
+        /// Helper class to pass on solution crawler events throttled with a certain delay.
+        /// arrive within a certain delay.
+        /// </summary>
+        private class SolutionCrawlerEventHandlerWithDelay
+        {
+            private readonly TimeSpan _publishInterval;
+            private readonly object _lock = new object();
+
+            private ProgressData _progressData;
+            private ProgressData _lastPublishedProgressData;
```

Comment:
it should never be necessary to remember the _lastPublishedProgressData.  All that we need is the last progress data we were *asked* to report.  if hte impl is correct, the UI will always reach a fixed point with that piece of data.